### DISCUSSION TITLE
Full incremental table reset/refresh

### DISCUSSION THREAD

**Discussion by 10xape** [TabularEditor/TabularEditor3#1413] [Data Modeling]
Hi I have added additional column to my dataflow table (incremental refresh) table that is linked to sementic model in Power BI service (also incremental refresh) and want to update it. Model is too large to do it from PB Desktop. I have tried to do full refresh of that table through TE3 but still cant see that additional column. I used Table full refresh, shoudl I select any other type of refreshing?

Thanks

Matt

**Comment by otykier** [comment]
A refresh operation does not update the *schema* of a table (i.e. columns and data types) - only the *data* in the columns that are currently specified on the table. To add the new column through TE3, you must either manually add it (right-click on the table, **Create** > **Data column**), or you can right click on the table and choose the [**Update table schema...**](https://docs.tabulareditor.com/te3/tutorials/importing-tables.html#updating-table-schema) option, to have TE3 attempt to update the schema automatically.

**Comment by 10xape** [comment]
thanks, I have tried right click update table and get thsi error: 
![image](https://github.com/user-attachments/assets/b0977ad4-9c96-41a5-a888-50f21a74edcf)

**Reply by otykier** [reply]
Could you share the M expression of your table? (Please blank out any sensitive information such as server names or credentials)

You may be able to work around this issue by enabled **Use Analysis Services for change detection** under **Tools > Preferences > Schema Compare**:
![image](https://github.com/user-attachments/assets/37554742-c589-4860-8219-2e515233a1de)

**Comment by 10xape** [comment]
Thank you, this is teh code:
```M
let
    // Function to enable incremental refresh
    SourceFunction = (RangeStart as nullable datetime, RangeEnd as nullable datetime) =>
    let
        // Original query steps
        Source = PowerPlatform.Dataflows(null),
        Workspaces = Source{[Id="Workspaces"]}[Data],
        #"XXXXXXXXXXXXXX" = Workspaces{[workspaceId="XXXXXXXXXXXXXX"]}[Data],
        #"XXXXXXXXXXXXXX" = #"XXXXXXXXXXXXXX"{[dataflowId="XXXXXXXXXXXXXX"]}[Data],
        cmc_cap_bridge_PQ2_ = #"XXXXXXXXXXXXXX"{[entity="cmc_cap_bridge_PQ2", version=""]}[Data],

        // Ensure 'refresh_date' is of type datetime
        ConvertedType = Table.TransformColumnTypes(cmc_cap_bridge_PQ2_, {{"refresh_date", type datetime}}),

        // Filter the data using RangeStart and RangeEnd
        FilteredRows = Table.SelectRows(
            ConvertedType,
            each [refresh_date] >= RangeStart and [refresh_date] < RangeEnd
        ),

        // Remove duplicates by 'unique_identifier'
        RemovedDuplicates = Table.Distinct(FilteredRows, {"unique_identifier"})
    in
        RemovedDuplicates,

    // Default parameter values for initial setup
    DefaultRangeStart = #datetime(2023, 1, 1, 0, 0, 0),
    DefaultRangeEnd = DateTimeZone.SwitchZone(DateTimeZone.FixedUtcNow(), 0),

    // Invoking the function with default or provided parameters
    InvokedFunction =
        if RangeStart = null and RangeEnd = null then
            SourceFunction(DefaultRangeStart, DefaultRangeEnd)
        else
            SourceFunction(RangeStart, RangeEnd)
in
    InvokedFunction
```
    
I am also tryint the other method that you advised about but it still has very long refreshing time.

**Reply by otykier** [reply]
Regarding the issue about the refresh taking a long time, I noticed that you have a `Table.Distinct` in the M query. This should generally be avoided, as this type of operation can be incredibly slow. Better to push that upstream to your dataflow, or better yet, to the source of the dataflow.

**Reply by 10xape** [reply]
I  think I already did it , I sometimes repat that twice to make sure it is not duplicating cuz of folding or some other actions, thank for teh tip

**Reply by otykier** [reply]
FWIW, we have a new update out now, which solves the **Unable to detect schema changes** error, that you reported initially: https://docs.tabulareditor.com/te3/other/release-notes/3_19_0.html?tabs=downloads

**Comment by 10xape** [comment]
Alos after scheme update I got the column and values just for one day (most recent) while in the source dataflow the value occurs accross the months, so I am not sure what is happenieng..... any assiatnce appriciated  thank you!  Matt

**Reply by otykier** [reply]
This sounds mostly like an issue with how the incremental refresh is configured. My recommendation is that you try to set up import of the table, including incremental refresh, in a new, temporary model in Power BI Desktop. Once you have that test model working, you can use Tabular Editor to copy over the table metadata to the model you're currently working on. That should ensure that the table is set up correctly.

**Reply by 10xape** [reply]
this is great idea ! is there any special techniques that I should follow while doing the copy of the metadata? It will be my first time doing that, would a few steps or instruction docs ref possible if you dont mind ?

**Reply by 10xape** [reply]
I mean coping metadata :)

**Reply by otykier** [reply]
It's as simple as opening two instances of Tabular Editor - connect one to Power BI Desktop with your temporary model, and the other instance to the model you're currently working on.

Then - in the first instance, copy (Ctrl+C) the table you just set up in Power BI Desktop. In the other instance, select the table you just recreated in Power BI Desktop, then hit paste (Ctrl+V). You should then see the following dialog to confirm that you will override the destination table with the one you just copied:

![image](https://github.com/user-attachments/assets/046c9a01-bda5-45c1-8f87-bca3dca6e887)

**Reply by 10xape** [reply]
oh so it is just a copying of the M code to the PB serv model, ok thanks

**Reply by 10xape** [reply]
Hey Daniel could you please unfold a little bit on that subject? I looked at this docs article https://docs.tabulareditor.com/te3/tutorials/incremental-refresh/incremental-refresh-workspace-mode.html   I have a problem with getting my had aroudn where the metadata is actually stored? Is that a source expression M code or I pull it from the other option/view? Could you please provide a full screen shots on how you access meetadat for the worksapce incremental refresh model in TE3? Thanks so much!

**Reply by otykier** [reply]
This thread is getting a bit unwieldy, and it's a bit unclear to me what your exact goal is. I would be happy to help, but may I ask that you post a new discussion/issue where you clearly state what exactly you are trying to achieve, what data source you are using, what you have done in Power BI, what you are trying to do with Tabular Editor 3, and the issues that you are seeing (error messages, etc.). Thank you!

**Reply by 10xape** [reply]
thanks and sorry for that, I understand some of these  are a beginner questions so I really appricate your patience. I did this thread yesterday that I think is going to cover it : Updating worspace incremental refresh model #1

**Comment by edhans** [comment]
Your code above will not work. The source data must be DateTime for IR to work, and the ChangeType you are doing will break folding. Get with your DBA to have the data converted to DateTime in the source. Then folding will happen and it should refresh normally.

What is happening with the long refresh is 100% of the table is being downloaded _for every partition_ and then it discards the data that isn't needed - data outside your RangeStart/RangeEnd date ranges for each partition. So if you have 12 monthly partitions, it is downloading the entire table 12 times. It would be faster to discard IR in this case and just refresh the table one time.

**Reply by 10xape** [reply]
Hey edhans, the time_lable column is not the data, it is a label for a time of the day. the date is in another column refresh_date :)

**Reply by edhans** [reply]
It doesn't matter. You are putting a Table.TransformColumnType() function in the query _before_ the incremental refresh filter. Comment it out or move it after the IR filter and see if that has an effect.

**Reply by 10xape** [reply]
you are right, thanks ! I am applying that now.

**Reply by 10xape** [reply]
I cannot update the code from TE2 editor, this error is poping up, this is for duplicates removal but both amendments have teh same error 
![image](https://github.com/user-attachments/assets/5a58c74a-2c26-477c-bea1-3908982c6405)

**Comment by otykier** [comment]
As mentioned in a comment above, we now have an update available which solves the issues mentioned in this thread, regarding performing the table schema update: https://docs.tabulareditor.com/te3/other/release-notes/3_19_0.html?tabs=downloads
